### PR TITLE
Definitions for new GeoFence, Rally Point spec

### PR DESCRIPTION
**Pull request by DonLakeFlyer**
This should give us enough to get started with new support. PX4 will implement and work through issues.

Discussion is here: #626

### CODE REVIEW

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -2624,12 +2682,16 @@
       <field type="float" name="x">PARAM5 / local: x position, global: latitude</field>
       <field type="float" name="y">PARAM6 / y position: global: longitude</field>
       <field type="float" name="z">PARAM7 / z position: global: altitude (relative or absolute, depending on frame.</field>
+      <extensions/>
+      <field type="uint8_t" name="mission_type">Mission type, see MAV_MISSION_TYPE</field>
```
Comment (WickedShell) [inline message_definitions/v1.0/common.xml]: Need to extend MISSION_ITEM_INT as well.

File: message_definitions/v1.0/common.xml [hunk 2]
```diff
@@ -1445,6 +1445,39 @@
...
+        <param index="1">Polygon vertex count</param>
+        <param index="2">Reserved</param>
+        <param index="3">Reserved</param>
+        <param index="4">Reserved</param>
+        <param index="5">Latitude</param>
+        <param index="6">Longitude</param>
+        <param index="7">Reserved</param>
+      </entry>
+      <entry value="5100" name="MAV_CMD_RALLY_POINT">
+        <description>Rally point. You can have multiple rally points defined.
```
Comment (julianoes) [inline message_definitions/v1.0/common.xml]: Can you please add an explanation what a rally point is. I keep thinking of a car rally race.

File: message_definitions/v1.0/common.xml [hunk 3]
```diff
@@ -2062,6 +2095,27 @@
...
+        <description>Autopilot supports mission rally point protocol.</description>
+      </entry>
+    </enum>
+    <enum name="MAV_MISSION_TYPE">
+      <description>Type of mission items being requested/sent in mission protocol.</description>
+      <entry value="0" name="MAV_MISSION_TYPE_MISSION">
+        <description>Items are mission commands for main mission.</description>
+      </entry>
+      <entry value="1" name="MAV_MISSION_TYPE_FENCE">
+        <description>Items are MAV_CMD_FENCE_ GeoFence items.</description>
```
Comment (julianoes) [inline message_definitions/v1.0/common.xml]: `GeoFence` -> `geofence` in my eyes.

**Comment by bkueng** [comment]
There are now 2 transactions needed for inclusion & exclusion polygons. How is the `MAV_CMD_NAV_FENCE_RETURN_POINT` handled? Can there be one for inclusion and one for exclusion? It might be simpler to only have one mission fence type and have `MAV_CMD_NAV_FENCE_POLYGON_VERTEX_INCLUSION`, `MAV_CMD_NAV_FENCE_POLYGON_VERTEX_EXCLUSION` instead. What do you think?

**Comment by DonLakeFlyer** [comment]
```MAV_CMD_NAV_FENCE_RETURN_POINT``` This is a return point which is used when the fence is breached. There is only one of them. If you look here: http://ardupilot.org/plane/docs/geofencing.html?highlight=fence. You can see how ArduPilot handles it.

With the addition of an exclusion fence I'm not really sure how you can make breach return work all the time. With a single breach return and a fence breach on exclusion fence the vehicle could easily fly through the exclusion zone to get to the breach return. I would guess when there is an exclusion fence breach the vehicle would not go to breach return but take some other action defined by a parameter.

I'm fine with doing inclusion and exclusion in single transaction.

@bkueng Can you take over updating mavlink and get this merged in so we can start implementing. Then we can adjust along the way. I'm hoping we are close enough now to begin work.

**Comment by DonLakeFlyer** [comment]
Superceded by #683

**Pull request closed without merging**